### DISCUSSION TITLE
Sword & Horde - Development Diary #2 - Room Appearance Layout

### DISCUSSION THREAD

**Discussion by RodRitter** [evennia/evennia#2669] [My Evennia Project]
This entry is about how we'll be able to see a room. What does the room look like? What's in the room? Where are the exits?

Here we'll see what the "look" command will reveal.

![image](https://user-images.githubusercontent.com/4159679/156940955-4e471767-5b70-4336-b9fa-5fd51bb2957a.png)

## Let's Build It
We'll need to use the default `Room` object, hooking into the `return_appearance` hook.
```
# rooms.py
class Room(DefaultRoom):
    def return_appearance(self):
        # Whatever we return here, will be displayed when "look" is called in a room
```
### Setting up some colors to use
```
    # We're using ANSI color scheme but you could easily replace with xterm 256 colors
    color_default = "|=t" 
    color_title = "|c"
    color_desc = "|=t"
    color_label = "|=m"
    color_char = "|y"
    color_object = "|g"
```
### Room name
we start an `appearance_string` that will be returned at the end
```
    appearance_string = "\n" # Breathing room after last command
    appearance_string += f"{color_title}{self.name}"
```
![image](https://user-images.githubusercontent.com/4159679/156941458-850a8c83-71c7-47b4-890a-f2acb0f2f027.png)

### Room Description
```
    if self.db.desc != None and self.db.desc.strip() != "":
        appearance_string += "\n"
        appearance_string += f"{color_desc}{self.db.desc}"
```
![image](https://user-images.githubusercontent.com/4159679/156941469-78b455b3-9430-4d32-bce6-d53c9cade287.png)

### List of Characters & Objects
```
    # filter for only objects that are Characters or plain old Objects
    objects_in_room = list(filter(self.is_char_or_obj, self.contents))

    # Only show "you see:" label if there is at least 1 object to show
    objects_in_room = list(filter(self.is_char_or_obj, self.contents))
    if len(objects_in_room) > 0:
        appearance_string += "\n"
        appearance_string += f"{color_label}You see: "
    
    for index, obj in enumerate(objects_in_room):
        # Characters are yellow, Objects are green
        color = color_char if "typeclasses.character" in obj.db_typeclass_path else color_object
        appearance_string += f"{color}{obj.name}"
        if index != len(objects_in_room)-1:
            appearance_string += f"{color_default}, "
```
Here is the filter that's used to get just Characters and Objects
```
    def is_char_or_obj(self, obj):
        path = obj.db_typeclass_path
        obj_type = "typeclasses.objects"
        char_type = "typeclasses.characters"
        return True if obj_type in path or char_type in path else False
```
![image](https://user-images.githubusercontent.com/4159679/156942213-5b875f13-2905-4e0d-bcdd-c5ec85a279be.png)

### The same thing for the list of Exits

```
    exits_in_room = list(filter(self.is_exit, self.contents))
    if len(exits_in_room) > 0:
        appearance_string += "\n"
        appearance_string += f"{color_label}exits: "
    
    for index, ext in enumerate(exits_in_room):
        appearance_string += f"{color_default}{ext.name}"
        if index != len(exits_in_room)-1:
            appearance_string += f"{color_default}, "
    return appearance_string
```
Filter:
```
    def is_exit(self, obj):
        return True if "typeclasses.exits" in obj.db_typeclass_path else False
```
![image](https://user-images.githubusercontent.com/4159679/156942289-045bb67e-954c-4aed-959b-b93ce411e7f1.png)

## Conclusion
This is all we need for now. When we start adding enemy NPC's, we'll modify this code to highlight them in red.

In the next entry, we'll be working on some **basic character stats** and start fleshing out the **NPC class**.

**Comment by jussolutions** [comment]
Hi Rod,

Thanks for doing this as my son has started to play around with this MUD engine to learn more about python. My background is mainly javascript so I prefer to write smaller scripts and I have been helping my son but we end up in a circular import, but I think the issue is that I am using separate files that call each other rather than sticking everything in rooms.py. What I have is the following:

**File: sewerbase.py**
```
from typeclasses.rooms import Room

class SewerBase( Room ):
    pass
```

**File: sewerentrance.py**
```
from world.sewer.sewerbase import SewerBase
from world.sewer.sewertunnel import SewerTunnel <----- ADDING THIS LINE GIVES ME A CIRCULAR IMPORT ISSUE

class SewerEntrance( SewerBase ):
    """
    The sewer entrance
    """
    def at_object_creation(self):
        self.db.desc = "The sewer entrance"
        create_object(DefaultExit,
                                  key="south",
                                  aliases=["s"],
                                  location=self,
                                  destination=SewerTunnel)
```

**File: sewertunnel.py**
```
from world.sewer.sewerbase import SewerBase

class SewerTunnel( SewerBase ):
    """
    The sewer entrance
    """
    def at_object_creation(self):
        self.db.desc = "In the sewer tunnel"
```

I think the circular import issue is due to the from SewerBase being imported multiple times, within SewerEntrance I have the SewerBase being imported then I import SewerTunnel which in turns import SewerBase again.

I think the resolution is to have all the room information in one file room.py but as his project expands and he adds 100 rooms with all the NPCs and items the file is going to become massive which is why i said potentially break them out into separate files.

Is this something that you have come across before, I have covered a lot of the documentation already but there is no recommendation on best practices for using the MUD, unless I have missed something, especially when it comes to creating rooms and exits.

**Reply by jaborsh** [reply]
Let me just say, I find it incredibly endearing that you and your son are working on a project together.

As for the issue, I don't believe that the problem you're encountering is due to a circular import. Rather, your issue is regarding the line `destination=SewerTunnel`. See, when you're defining classes within Evennia, you are not creating objects. You are simply designing a framework (see: `typeclasses`) about how a potential object should function.

You're essentially saying that you want the destination from your Sewer Entrance to be a framework, but obviously that doesn't make much sense. You'll need to have your Sewer Tunnel created so that it can be queried in the database. Destination accepts a database object, not a typeclass.

**Reply by Tegiminis** [reply]
There's a careful balance to be struck between breaking everything into individual files and keeping it all in one file.

If you put everything into individual files, you are going to get a lot of circular import errors like this, and your project's file count is going to balloon. You mention adding 100 unique rooms and making them all individual files, but that now means you have... 100 individual files. Which is even more annoying to comb through when looking for a bug, especially the circular imports that are likely to plague such a project.

Conversely, you definitely identified the problem with putting everything in one big file. Then you have the opposite problem; you know where stuff is and there's no circular import problems, but the line count is in the thousands.

My suggestion is to identify discrete chunks of information you can bundle up, and then put them into unique files. For example, these three files could probably be just one: `sewerbase.py`. This removes your import problems and neatly organizes your code into understandable chunks.

---

However, I would like to propose a different solution altogether. You are writing code for this when you don't have to.

Think of Evennia as multiple layers. You have the code layer, where all your methods and functions and object references live, but you also have the database layer, which holds all the instantiated objects for future use.

Creating a new room with no extra functionality doesn't require any code at all. It just requires you instantiate a new room object with the data you want. This is easily handled by the existing evennia commands `tunnel` and `dig`. [Here's a tutorial](https://www.evennia.com/docs/latest/Building-Quickstart.html?highlight=tunnel) on the basics of how you do that in-client.

Should you want to make the room or exits have unique functionality - like, say, the Sewer periodically floods with wastewater, or passing through the exit causes a debuff - then you will need to extend `Room` into `SewerRoom`. But if you're just changing descriptions, creating exits, etc, you're much better off using the in-client builder tools. And when you do create SewerRoom with unique functionality, you'll still be creating it in the database using the in-client tools 99% of the time.

Hope this helps!

**Reply by jussolutions** [reply]
> Let me just say, I find it incredibly endearing that you and your son are working on a project together.
> 
> As for the issue, I don't believe that the problem you're encountering is due to a circular import. Rather, your issue is regarding the line `destination=SewerTunnel`. See, when you're defining classes within Evennia, you are not creating objects. You are simply designing a framework (see: `typeclasses`) about how a potential object should function.
> 
> You're essentially saying that you want the destination from your Sewer Entrance to be a framework, but obviously that doesn't make much sense. You'll need to have your Sewer Tunnel created so that it can be queried in the database. Destination accepts a database object, not a typeclass.

@The-Archemist

Bless him as he loves developing, he is only 12 and learnt python about 6 months to go to write some bots in discord for Roblox but wants to branch out, so thought I would help as MUDs was one of the first things I ever played with so has some nostalgia feelings for me from Uni :-)

On the answer, what a dummy I am, you are absolutely right, I have defined the class but not created it yet. Thank you for this.

**Reply by jussolutions** [reply]
> There's a careful balance to be struck between breaking everything into individual files and keeping it all in one file.
> 
> If you put everything into individual files, you are going to get a lot of circular import errors like this, and your project's file count is going to balloon. You mention adding 100 unique rooms and making them all individual files, but that now means you have... 100 individual files. Which is even more annoying to comb through when looking for a bug, especially the circular imports that are likely to plague such a project.
> 
> Conversely, you definitely identified the problem with putting everything in one big file. Then you have the opposite problem; you know where stuff is and there's no circular import problems, but the line count is in the thousands.
> 
> My suggestion is to identify discrete chunks of information you can bundle up, and then put them into unique files. For example, these three files could probably be just one: `sewerbase.py`. This removes your import problems and neatly organizes your code into understandable chunks.
> 
> However, I would like to propose a different solution altogether. You are writing code for this when you don't have to.
> 
> Think of Evennia as multiple layers. You have the code layer, where all your methods and functions and object references live, but you also have the database layer, which holds all the instantiated objects for future use.
> 
> Creating a new room with no extra functionality doesn't require any code at all. It just requires you instantiate a new room object with the data you want. This is easily handled by the existing evennia commands `tunnel` and `dig`. [Here's a tutorial](https://www.evennia.com/docs/latest/Building-Quickstart.html?highlight=tunnel) on the basics of how you do that in-client.
> 
> Should you want to make the room or exits have unique functionality - like, say, the Sewer periodically floods with wastewater, or passing through the exit causes a debuff - then you will need to extend `Room` into `SewerRoom`. But if you're just changing descriptions, creating exits, etc, you're much better off using the in-client builder tools. And when you do create SewerRoom with unique functionality, you'll still be creating it in the database using the in-client tools 99% of the time.
> 
> Hope this helps!

@Tegiminis

I defiantly can see you view and the circular imports is frustrating but it is something that I think as you have said maybe going into a single file for the rooms for the areas might be better which is what I thought last night, and is something that I will get him to do.

I did look at the dig command but the goal of the exercise for him is to learn python more so the dig command would not give him that level of learning but again we do use that command to generate the areas once we have defined them. The sewer water is something that we have spoken about which is why we want to make sure we have the sewer base class, for example this area is going to be by default in the dark so through inheritance we can set the default behaviour of the room, we can also use tags as well to be able to identify the areas which we will do as well.

Really good fun though.

**Reply by RodRitter** [reply]
@jussolutions Looks like you've gotten a lot of feedback from the devs. I'm also a JS-focused dev and have only just gotten familiar with Python and Evennia. I've gone through a few hell-dives for problems similar to this. So if you need a sounding board, let me know and I'm happy to connect on Discord or somehwere.

Regarding your issue, if it helps, I think the biggest thing you could understand is the difference between **classes** and a **class instance**. You could potentially create many Sewer Entrances, so the Exit destination needs a specific instance of the Entrance, not the class.

I read somewhere below that you're trying not to use the @dig/@tunnel command, but that handles the instancing issue you're facing. Perhaps you can start by building your own "tunnel" command. That should help you understand it better.

CustomTunnel Command: _You start inside an existing room already (ie. SewerEntrance)_.
1. Set up the command. Make sure you set permissions to builder so players can't tunnel:
```
    key = "tunnel"
    aliases = ["tun"]
    lock = "cmd:perm(Builders)"
    help_category = "Builders"
```

2. Create a new SewerTunnel Room - 
```
new_room = create_object(typeclass=SewerTunnel, key="Sewer Tunnel", home=self.caller.location)
```

3. Let's connect the rooms with Exits:

```
entrance_to_tunnel = create_object(typeclass=Exit, key="Rusty Grate", aliases=["rusty", "grate"], location=self.caller.location, destination=new_room)

tunnel_to_entrance = create_object(typeclass=Exit, key="Rusty Grate", aliases=["rusty", "grate"], location=new_room, destination=self.caller.location)
```

and that's really all there is to the major logic. This is spuedo code, so you'll need to add some parsing of `self.args`, but it should give you what you need.

**Comment by jussolutions** [comment]
Hi,

So I have taken all the advice and we consolidated into a single file but we are still getting errors, I think I might add some information to the Exit help file as it is not really clear how to create rooms and exits within Python and searching on the internet provides very little information, unless I am searching for the wrong thing.

Anyway, what we have done is consolidate the files into one, please see below:

```
from typeclasses.rooms import Room
from evennia.utils.search import search_object
from evennia import create_object, DefaultExit


################################################################################
#
# The base sewer room
#
################################################################################
class SewerBaseRoom( Room ):
    """
    This room class is used for all character-generation rooms. It makes the
    chargencmdset available
    """

    pass

################################################################################
#
# The Sewer Entrance
#
################################################################################
class SewerEntrance( SewerBaseRoom ):
    """
    This is the sewer entrance
    """
    def at_object_creation(self):
        self.db.desc = "You have entered the sewers, why you want to be here, only you know. It smells.\n" \
                       "There is a number of lit torches running down one wall that makes this entrance\n" \
                       "very bright, the last torch is unlit and looks loose."


        sewertunnel1 = search_object( "sewertunnel1" )
        if not sewertunnel1:
            sewertunnel1 = create_object(SewerTunnel1, key="sewertunnel1", location=None)
        create_object(DefaultExit, key="south", aliases=["s"], location=self, destination=sewertunnel1)

################################################################################
#
# Sewer Tunnel 1
#
################################################################################
class SewerTunnel1( SewerBaseRoom ):
    """
    This will be the first tunnel 
    """
    def at_object_creation(self):
        self.db.desc = "You have started exploring the sewer tunnels, you have a strange feeling that\n" \
                       "creatures lurk around these dark rooms"

        sewerentrance = search_object( "sewerentrance" )
      
        if not sewerentrance:
            sewerentrance = create_object( SewerEntrance, key="sewerentrance", location=None)
      
        create_object(DefaultExit, key="north", aliases=["n"], location=self, destination=sewerentrance)
```
There is some work that I need to move on the movement, which I have seen some documentation for as north, south, east and west is more like commands rather than exits but I thought the above would work.

When I run this command I get the following error:

ValueError: Cannot assign "<QuerySet [sewerentrance]>": "DefaultExit.db_destination" must be a
"ObjectDB" instance.

I did try it with no search and just the create_object but I get the following error,

RecursionError: maximum recursion depth exceeded while calling a Python object

Again the recursive issue comes back. I am on the right track when it comes to exits within rooms?

Thanks again for your help.

**Reply by aMiss-aWry** [reply]
search_object("sewerentrance") looks for an item by name of sewerentrance. You don't have any such item, but you do have one called 'sewertunnel1'. 

Replace 
search_object("sewerentrance") 

with 
search_object("sewertunnel1")

**Reply by aMiss-aWry** [reply]
Sorry, my mistake - the problem is actually that you have two classes looping heh.

The logic is:

In SewerEntrance
If there's no sewer tunnel, make a sewer tunnel

In SewerTunnel
If there's no sewer entrance, make a sewer entrance

So it looks for a sewer tunnel, doesn't find one, looks for a sewer entrance, doesn't find one, looks for a sewer tunnel, doesn't find one, looks for a sewer entrance, doesn't find one... and it never finishes the generating code, hence the recursion.

**Reply by aMiss-aWry** [reply]
Since the search is responding with a queryset you could try destination=sewerentrance[0] and destination=sewertunnel1[0] instead. It's probably returning the search results in a set.

**Reply by jussolutions** [reply]
Hi,

So what I thought might be happening is that I use the following:

dig sewerentrance:world.sewers.SewerEntrance = east

That creates the initial sewer entrance, so when I then go through that class it should do the following:

```
        sewertunnel1 = search_object( "sewertunnel1" )
        if not sewertunnel1:
            sewertunnel1 = create_object(SewerTunnel1, key="sewertunnel1", location=None)
```
So sewertunnel1 will be None to start with, but then the create_object of the if statement will execute to create the object and store it in sewertunnel1.

Then the create_object on the exit will run. In sewertunnel the sewerentrance is already created by the dig command I can see the entries in the database.

It could be I have different types, so I will try changing the if statement so that if I create the object I can add the exit using sewertunnel1 else i will need to use sewertunnel1[0] as a query set is returned. I guess!
